### DISCUSSION TITLE
What on earth does this code do?

### DISCUSSION THREAD

**Discussion by axkibe** [libgit2/libgit2#6174] [Q&A] [answered]
Found at https://github.com/libgit2/libgit2/blob/main/src/util.h#L29

```#if defined(__GNUC__)
# define GIT_CONTAINER_OF(ptr, type, member) \
    __builtin_choose_expr( \
        __builtin_offsetof(type, member) == 0 && \
        __builtin_types_compatible_p(__typeof__(&((type *) 0)->member), __typeof__(ptr)), \
        ((type *) (ptr)), \
        (void)0)
#else
# define GIT_CONTAINER_OF(ptr, type, member) (type *)(ptr)
#endif
```
That macro is used quite often, is this some kind of upcast? (with an optional optimization for GNUC?) 

I'm really confused

**Comment by boretrk** [comment, marked as answer]
Not really optimization, it just adds some validation when compiling for GNUC to catch some of the possible miscasts.
It's for downcasting rather than upcasting.

Say that you have 
```
typedef struct
{
	enum fruittype;
} fruit;

typedef struct
{
	fruit parent;
} apple;

fruit *fruitptr;
```
If you then use `GIT_CONTAINER_OF(fruitptr, apple, parent)` to cast `fruitptr` to `apple` the macro will do compile time verification that `apple` actually starts with a `fruit` member named `parent`.
So if you accidentally add members to `apple` before `parent` or try to cast something that isn't a `fruit`-pointer to `apple`, GCC will complain.
If you don't use GCC it will just fall back to a regular cast without the validation.

**Reply by axkibe** [reply]
Oh thank you, now I understand.

Might we add a comment in the source code to explain it a bit?

IMO this code is all but obvious / self explanatory.
